plugins/linux-pipewire/pipewire.c

```diff
@@ -338,63 +341,65 @@ static inline struct spa_pod *build_format(obs_pipewire_stream *obs_pw_stream,
 	 * acceptable. The id gives more context about what the object is meant to
 	 * contain. In this case we enumerate supported formats. */
-	spa_pod_builder_push_object(b, &format_frame, SPA_TYPE_OBJECT_Format,
+	spa_pod_builder_push_object(&pod_builder.b, &format_frame,
+				    SPA_TYPE_OBJECT_Format,
 				    SPA_PARAM_EnumFormat);
 	/* add media type and media subtype properties */
-	spa_pod_builder_add(b, SPA_FORMAT_mediaType,
+	spa_pod_builder_add(&pod_builder.b, SPA_FORMAT_mediaType,
 			    SPA_POD_Id(SPA_MEDIA_TYPE_video), 0);
-	spa_pod_builder_add(b, SPA_FORMAT_mediaSubtype,
+	spa_pod_builder_add(&pod_builder.b, SPA_FORMAT_mediaSubtype,
 			    SPA_POD_Id(SPA_MEDIA_SUBTYPE_raw), 0);
 
 	/* formats */
-	spa_pod_builder_add(b, SPA_FORMAT_VIDEO_format, SPA_POD_Id(format), 0);
+	spa_pod_builder_add(&pod_builder.b, SPA_FORMAT_VIDEO_format,
+			    SPA_POD_Id(format), 0);
 
 	/* modifier */
 	if (modifier_count > 0) {
 		struct spa_pod_frame modifier_frame;
 
 		/* build an enumeration of modifiers */
-		spa_pod_builder_prop(b, SPA_FORMAT_VIDEO_modifier,
+		spa_pod_builder_prop(&pod_builder.b, SPA_FORMAT_VIDEO_modifier,
 				     SPA_POD_PROP_FLAG_MANDATORY |
 					     SPA_POD_PROP_FLAG_DONT_FIXATE);
 
-		spa_pod_builder_push_choice(b, &modifier_frame, SPA_CHOICE_Enum,
-					    0);
+		spa_pod_builder_push_choice(&pod_builder.b, &modifier_frame,
+					    SPA_CHOICE_Enum, 0);
 
 		/* The first element of choice pods is the preferred value. Here
 		 * we arbitrarily pick the first modifier as the preferred one.
 		 */
-		spa_pod_builder_long(b, modifiers[0]);
+		spa_pod_builder_long(&pod_builder.b, modifiers[0]);
 
 		/* modifiers from  an array */
 		for (uint32_t i = 0; i < modifier_count; i++)
-			spa_pod_builder_long(b, modifiers[i]);
+			spa_pod_builder_long(&pod_builder.b, modifiers[i]);
 
-		spa_pod_builder_pop(b, &modifier_frame);
+		spa_pod_builder_pop(&pod_builder.b, &modifier_frame);
 	}
 	/* add size and framerate ranges */
 	spa_pod_builder_add(
-		b, SPA_FORMAT_VIDEO_size,
+		&pod_builder.b, SPA_FORMAT_VIDEO_size,
 		SPA_POD_CHOICE_RANGE_Rectangle(&resolution, &min_resolution,
 					       &max_resolution),
 		SPA_FORMAT_VIDEO_framerate,
 		SPA_POD_CHOICE_RANGE_Fraction(&framerate, &min_framerate,
 					      &max_framerate),
 		0);
-	return spa_pod_builder_pop(b, &format_frame);
+	return spa_pod_builder_pop(&pod_builder.b, &format_frame);
 }
 
 static bool build_format_params(obs_pipewire_stream *obs_pw_stream,
-				struct spa_pod_builder *pod_builder,
-				const struct spa_pod ***param_list,
-				uint32_t *n_params)
+				GPtrArray **out_params)
 {
 	obs_pipewire *obs_pw = obs_pw_stream->obs_pw;
-	uint32_t params_count = 0;
+	g_autoptr(GPtrArray) params = NULL;
+
+	assert(out_params != NULL);
 
-	const struct spa_pod **params;
-	params = bzalloc(2 * obs_pw_stream->format_info.num *
-			 sizeof(struct spa_pod *));
+	params = g_ptr_array_new_full(2 * obs_pw_stream->format_info.num *
```

Comment:
If we want to use a dynamic array we should use the obs DARRAY to keep consistency with the rest of obs.

Reply:
DARRAY in this kind of code path (not cold) should be avoided.

Reply:
First of all negotiation is cold. And second we know the size so there arnt any performance concerns to begin with. Its just a matter of using the api we all know or using the gptr api.